src/arithmetic/arithmetic_expression.c

```diff
@@ -1178,6 +1149,169 @@ SIValue AR_NE(SIValue *argv, int argc) {
+// if indices are still integers but not in the valid range, only values within the valid range
+// will be returned
+// if one of the indices is null, null will be returnd
+// "RETURN [1, 2, 3][0..1]" will yield [1, 2]
+SIValue AR_SLICE(SIValue *argv, int argc) {
+	assert(argc == 3 && argv[0].type == T_ARRAY);
+	if(argv[0].type == T_NULL || argv[1].type == T_NULL || argv[2].type == T_NULL) return SI_NullVal();
+	assert(argv[1].type == T_INT64 && argv[2].type == T_INT64);
+	SIValue array = argv[0];
+
+	// get array length
+	uint32_t arrayLen = SIArray_Length(array);
+
+	// get start and end index
+	SIValue start = argv[1];
+	int32_t startIndex = (int32_t)start.longval;
+	SIValue end = argv[2];
+	int32_t endIndex = (int32_t)end.longval;
+
+	uint32_t absStartIndex = abs(startIndex);
+	// if negative index, calculate offset from end
+	startIndex = startIndex >= 0 ? startIndex : arrayLen - absStartIndex;
+	// if offset from the end is out of bound, start at 0
+	startIndex = startIndex < 0 ? 0 : startIndex;
+
+	uint32_t absEndIndex = abs(endIndex);
+	// if negative index, calculate offset from end
+	endIndex = endIndex >= 0 ? endIndex : arrayLen - absEndIndex;
+	// if index out of bound, end at arrayLen
+	endIndex = ((int32_t)arrayLen) < endIndex ? arrayLen : endIndex;
+	// cant go in reverse
+	if(endIndex <= startIndex) {
+		return SI_EmptyArray();
+	}
+
+	SIValue subArray = SI_Array(endIndex - startIndex);
+	for(uint i = startIndex; i < endIndex; i++) {
+		SIArray_Append(&subArray, SIArray_Get(array, i));
+	}
+	return subArray;
+}
+
+// create a new list of integers in the range of [start, end]. If an interval was given
+// the interval between two consecutive list members will be this interval.
+// If interval was not suppllied, it will be default as 1
+// "RETURN range(3,8,2)" will yield [3, 5, 7]
+SIValue AR_RANGE(SIValue *argv, int argc) {
+	assert(argc > 1 && argc <= 3 && argv[0].type == T_INT64 && argv[1].type == T_INT64);
+	int64_t start = argv[0].longval;
+	int64_t end = argv[1].longval;
+	int64_t interval = 1;
+	if(argc == 3) {
+		assert(argv[2].type == T_INT64);
+		interval = argv[2].longval;
+	}
+	SIValue array = SI_Array(1 + (end - start) / interval);
+	for(; start <= end; start += interval) {
+		SIArray_Append(&array, SI_LongVal(start));
+	}
+	return array;
+}
+
+// checks if a value is in a given list
+// "RETURN 3 IN [1, 2, 3]" will return true
+SIValue AR_IN(SIValue *argv, int argc) {
+	assert(argc == 2 && argv[1].type == T_ARRAY);
+	SIValue lookupValue = argv[0];
+	SIValue lookupList = argv[1];
+	bool comparedNull = false;
+	uint arrayLen = SIArray_Length(lookupList);
+	for(uint i = 0; i < arrayLen; i++) {
+		int compareValue = SIValue_Compare(lookupValue, SIArray_Get(lookupList, i));
+		if(compareValue == 0) return SI_BoolVal(true);
+		if(compareValue == COMPARED_NULL) comparedNull = true;
+	}
+	return comparedNull ? SI_NullVal() : SI_BoolVal(false);
```

Comment:
Maybe add a comment describing NULL-handling and the intended return!

Reply:
done